src/web3/KeychainSigner.ts

```diff
@@ -0,0 +1,203 @@
+  return CryptoJS.AES.encrypt(privateKey, password).toString()
+}
+
+async function decryptPrivateKey(encryptedPrivateKey: string, password: string) {
+  const bytes = CryptoJS.AES.decrypt(encryptedPrivateKey, password)
+  return bytes.toString(CryptoJS.enc.Utf8)
+}
+
+async function storePrivateKey(privateKey: string, account: KeychainAccount, password: string) {
+  const encryptedPrivateKey = await encryptPrivateKey(privateKey, password)
+  return storeItem({ key: accountStorageKey(account), value: encryptedPrivateKey })
+}
+
+async function getStoredPrivateKey(
+  account: KeychainAccount,
+  password: string
+): Promise<string | null> {
+  try {
+    Logger.debug(TAG, `Checking keystore for private key for account ${account}`)
+    const encryptedPrivateKey = await retrieveStoredItem(accountStorageKey(account))
+    if (!encryptedPrivateKey) {
+      throw new Error('No private key found in storage')
+    }
+
+    return await decryptPrivateKey(encryptedPrivateKey, password)
+  } catch (error) {
+    Logger.error(TAG, 'Failed to retrieve private key', error)
+    return null
+  }
+}
+
+// Returns accounts that have been stored in the keychain, sorted by creation date
+// The ordering is important: Geth KeyStore does this and some parts of the code base rely on it.
+export async function listStoredAccounts() {
+  let accounts: KeychainAccount[]
+  try {
+    const storedItems = await listStoredItems()
+    Logger.info(`${TAG}@listStoredAccounts`, 'Keychain items:', storedItems)
+    accounts = storedItems
+      .filter((item) => item.startsWith(ACCOUNT_STORAGE_KEY_PREFIX))
+      .map((item) => {
+        const [isoDate, rawAddress] = item.slice(ACCOUNT_STORAGE_KEY_PREFIX.length).split('--')
+        const address = normalizeAddressWith0x(rawAddress)
+        const createdAt = new Date(isoDate)
+        return {
+          address,
+          createdAt,
+        }
+      })
+      .sort((a, b) => a.createdAt.getTime() - b.createdAt.getTime())
+  } catch (e) {
+    Logger.error(`${TAG}@listStoredAccounts`, 'Error listing accounts', e)
+    throw new Error(ErrorMessages.GETH_FETCH_ACCOUNTS)
+  }
+  return accounts
+}
+
+/**
+ * Implements the signer interface on top of the OS keychain
+ */
+export class KeychainSigner implements Signer {
+  protected unlockedLocalSigner: LocalSigner | null = null
+  // Timestamp in milliseconds when the signer was last unlocked
+  protected unlockTime?: number
+  // Number of seconds that the signer was last unlocked for
+  protected unlockDuration?: number
+
+  /**
+   * Construct a new instance of the Keychain signer
+   *
+   * @param account Account address derived from the private key to be called in init
+   */
+  constructor(protected account: KeychainAccount) {}
+
+  async init(privateKey: string, passphrase: string) {
+    await storePrivateKey(privateKey, this.account, passphrase)
+  }
+
+  async signTransaction(
+    addToV: number,
```

Comment:
What does `addToV` mean?

Reply:
Good question:
> represents the chainId and is added to the recoveryId to prevent replay

https://github.com/celo-org/celo-monorepo/blob/master/packages/sdk/connect/src/wallet.ts#L18